dialer.go

```diff
@@ -148,6 +185,21 @@ func (d *dialer) BoundTo(dial DialFN) BoundDialer {
 	return &boundDialer{d, dial}
 }
 
+func (d *dialer) initSession(current eventual.Value, dial DialFN) {
+	s, err := d.startSession(dial)
+	if err != nil {
+		d.mx.Lock()
+		d.clearSession()
+		d.mx.Unlock()
+		current.SetIfEmpty(&sessionAndError{nil, err})
+		return
+	}
+	if !current.SetIfEmpty(&sessionAndError{s, nil}) {
```

Comment:
We can have multiple of these calls to initSession going at the same time. By using `SetIfEmpty` we ensure that only the first one is used. Other successfully opened sessions are closed immediately to avoid using resources.